File: 백준/Silver/1991. 트리 순회/트리 순회.cc
#include <iostream>

using std::cin;
using std::cout;

struct node_t{
    int left, right;
}arr[26];

void preorder(int curr) {
    cout << char(curr + 65);
    if(arr[curr].left > 0) preorder(arr[curr].left);
    if(arr[curr].right > 0) preorder(arr[curr].right);
}

void postorder(int curr) {
    if(arr[curr].left > 0) postorder(arr[curr].left);
    if(arr[curr].right > 0) postorder(arr[curr].right);
    cout << char(curr + 65);
}

void inorder(int curr) {
    if(arr[curr].left > 0) inorder(arr[curr].left);
    cout << char(curr + 65);
    if(arr[curr].right > 0) inorder(arr[curr].right);
}

int main() {
    int n; cin >> n;
    for(int i = 0; i < n; ++i) {
        char a,b,c;
        cin >> a >> b >> c;
        arr[a-65].left = b != '.' ? b-'A' : -1;
        arr[a-65].right = c != '.' ? c-'A' : -1;
    }

    preorder(0); puts("");
    inorder(0); puts("");
    postorder(0); puts("");
    return 0;
}

File: 백준/Silver/1991. 트리 순회/트리 순회.java
import java.io.IOException;
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.StringTokenizer;

public class Main {
    static char[] left = new char[28], right = new char[28];
    static int N;
    static StringBuilder sb = new StringBuilder();
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        N = Integer.parseInt(br.readLine());
        for(int i = 0; i < N; ++i) {
            StringTokenizer st = new StringTokenizer(br.readLine(), " ");
            char P = st.nextToken().charAt(0), L = st.nextToken().charAt(0), R = st.nextToken().charAt(0);
            P = (char)(P - 'A' + 1);
            L = L == '.' ? (char)0 : (char)(L - 'A' + 1);
            R = R == '.' ? (char)0 : (char)(R - 'A' + 1);
            left[P] = L;
            right[P] = R;
        }

        preOrder(1); sb.append('\n');
        inOrder(1); sb.append('\n');
        postOrder(1);
        System.out.println(sb);
    }

    static void preOrder(int curr) {
        sb.append((char)(curr + 'A' - 1));
        if (left[curr] != 0) preOrder(left[curr]);
        if (right[curr] != 0) preOrder(right[curr]);
    }

    static void inOrder(int curr) {
        if (left[curr] != 0) inOrder(left[curr]);
        sb.append((char)(curr + 'A' - 1));
        if (right[curr] != 0) inOrder(right[curr]);
    }

    static void postOrder(int curr) {
        if (left[curr] != 0) postOrder(left[curr]);
        if (right[curr] != 0) postOrder(right[curr]);
        sb.append((char)(curr + 'A' - 1));
    }
}